c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+    C10_CUDA_CHECK(cudaDeviceGetDefaultMemPool(&mempool, device));
+    uint64_t threshold = UINT64_MAX;
+    C10_CUDA_CHECK(cudaMemPoolSetAttribute(mempool, cudaMemPoolAttrReleaseThreshold, &threshold));
+
+    // I think all these are on by default, but I want to enable them
+    // explicitly to ensure awareness.
+    int enable = 1;
+    C10_CUDA_CHECK(cudaMemPoolSetAttribute(mempool,
+                                           cudaMemPoolReuseFollowEventDependencies,
+                                           &enable));
+    C10_CUDA_CHECK(cudaMemPoolSetAttribute(mempool,
+                                           cudaMemPoolReuseAllowOpportunistic,
+                                           &enable));
+    C10_CUDA_CHECK(cudaMemPoolSetAttribute(mempool,
+                                           cudaMemPoolReuseAllowInternalDependencies,
+                                           &enable));
+
+    // Grabs a stream from the current device to use as the "unifier" free stream
+    // for allocations that end up used on multiple streams.
+    const auto dufs = getStreamFromPool();
+    dummy_unifying_free_streams[device] = UsageStream(dufs.stream(), dufs.device_index());
+
+    pytorch_used_bytes[device] = 0;
+    pytorch_memory_limits[device] = UINT64_MAX;
+
+    devs_initialized_flags[device] = true;
+  }
+}
+
+// Assumes the caller holds general_mutex
+inline void free_impl(PtrInfo::iterator& it) {
+  // Possible micro-optimization: If we did a value-copy here, we could move
+  // ptr_info.erase(it) up here and drop the lock immediately.
+  const auto& usage_streams = it->second.usage_streams;
+
+  // If the usage stream is a null (default) stream,
+  // cudaFreeAsync infers the device from the ambient context,
+  // so we need to set the right ambient context.
+  CUDAGuard g(usage_streams[0].device);
+
+  if (usage_streams.size() == 1) {
+    // ptr was only used on one stream, which must have been
+    // the original allocation stream.
+    // Frees ptr in the original allocation stream.
+    C10_CUDA_CHECK(cudaFreeAsync(it->first, usage_streams[0].stream));
+
+    if (C10_UNLIKELY(capture_underway)) {
+      // See Note [Avoid dangling free streams during CUDA graph capture]
+      capture_free_streams.insert(usage_streams[0]);
+    }
+  } else {
+    // ptr was used on many streams. We don't know which was the most recent.
+    // There could even have been multiple most recent usage streams acting
+    // on different regions of the memory.
+    // But cudaFreeAsync only accepts a single most recent usage stream.
+    // We can still safely free ptr with a trick:
+    // Use a dummy "unifying stream", sync the unifying stream with all of
+    // ptr's usage streams, and pass the dummy stream to cudaFreeAsync.
+
+    // Retrieves the dummy "unifier" stream from the device
+    // on which the pointer was originally allocated.
+    auto dummy_unifying_free_stream = dummy_unifying_free_streams[usage_streams[0].device];
```

Comment:
Is it necessary for there to be a single dummy unifying stream per device? It seems to me that you could just fetch a stream (any stream) and use it to do the unification. The behavior here seems to serialize more than is necessary.

Reply:
That's not a bad idea and I think it would be correct code, but two things are at risk of false serialization here: memory reuse and stream work.

If, when freeing every used-on-several-streams allocation, we pull a new random stream and use it as the unifying free stream, the random stream will be told to wait for the usage streams, which could spuriously delay work enqueued on the random stream afterward.

Hmm, I guess there's still a danger of this if a dummy_unifying_free_stream happens to alias with someone's work stream. To fully avoid false work serialization we'd need dummy_unifying_free_stream to be unique to the mempool. Maybe I should bypass the stream pool and use a raw cudaStreamCreate.

Reply:
I thought getting it from cudaStreamCreate doesn't save you from false dependencies, as all streams are still mapped onto finite number of hardware streams? Also, we've had cases where cudaStreamCreates/Destroys (or maybe cudaEventCreates/Destroys) lead to distributed hangs for some very non-obvious and poorly diagnosable reasons?